syntax = "proto3";

message no_parameter {}

message request_limiter {
  int64 limit = 1;
}

message base_operation_result {
  bool success = 1;
  string error_message = 2;
}

message new_sensor_reading_save_request {
  string sensor_id = 1;
  string reading_location_id = 2;
  float reading_value = 3;
  string read_at = 4;
}

message new_sensor_reading_save_response {
  string reading_id = 1;
  base_operation_result result = 2;
}

message sensor_reading_item {
  string id = 1;
  string sensor_id = 2;
  string reading_location_id = 3;
  float reading_value = 4;
  bool trusted_reading = 5;
  bool location_known = 6;
  bool location_accurate = 7;
  bool is_processed = 8;
  string read_at = 9;
  string updated_at = 10;
  string received_at = 11;
  float overall_integrity = 12;
}

message sensor_reading_fetch_single_item_request {
  string reading_id = 1;
}

message sensor_reading_fetch_single_item_response {
  sensor_reading_item item = 1;
  base_operation_result result = 2;
}

message sensor_reading_fetch_multi_item_response {
  repeated sensor_reading_item items = 1;
  double item_count = 2;
  base_operation_result result = 3;
}

service SensorService {
  rpc save_reading(new_sensor_reading_save_request) returns (new_sensor_reading_save_response) {}
  rpc get_readings(request_limiter) returns (sensor_reading_fetch_multi_item_response) {}
  rpc get_reading(sensor_reading_fetch_single_item_request) returns (sensor_reading_fetch_single_item_response) {}
}
